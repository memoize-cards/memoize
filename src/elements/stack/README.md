# Stack

The `stack` component is a flexible and versatile styled component that utilizes flexbox properties for layout management. It provides an easy way to create responsive and dynamic layouts with alignment, content positioning, direction, and gap customization.

### Usage

1. Import the `stack` component in your project.

```js
import stack from '@elements/stack'
```

2. Use the `stack` component in your code to create flexible layouts.

```js
import stack from '@elements/stack'

const App = () => {
  return (
    <stack.Div column center>
      <div>Item 1</div>
      <div>Item 2</div>
      <div>Item 3</div>
    </stack.Div>
  )
}

export default App
```

### Contributing

Contributions are welcome! If you find any issues or have suggestions for improvements, please open an issue or submit a pull request on the GitHub repository for this package.

### License

This package is licensed under the [MIT License](https://opensource.org/licenses/MIT).
