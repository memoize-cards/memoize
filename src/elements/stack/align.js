/**
 * @typedef {'top'|'middle'|'bottom'} Alignments - One of the valid align values.
 */
const alignments = [
  'top',
  'middle',
  'bottom'
]

/**
 * Maps alignment values to alignment keywords.
 *
 * @type {Object.<Alignments, string>}
 */
const mapper = {
  top: 'start',
  middle: 'center',
  bottom: 'end'
}

/**
 * Returns the alignment keyword based on the provided properties.
 *
 * @param {Object.<Alignments, boolean>} props - The component properties containing the alignment values.
 * @returns {string} The calculated alignment. One of: 'top', 'middle', 'bottom', or 'top' (default).
 */
function align (props) {
  return mapper[alignments.find((key) => props[key])] ?? mapper.top
}

export default align
