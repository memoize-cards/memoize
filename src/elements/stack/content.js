const alignments = [
  'left',
  'center',
  'right'
]

const mapper = {
  left: 'start',
  center: 'center',
  right: 'end'
}

function content (props) {
  return mapper[alignments.find((key) => props[key])] ?? mapper.left
}

export default content
