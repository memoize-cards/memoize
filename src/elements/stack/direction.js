const directions = [
  'row',
  'column'
]

function direction (props) {
  return directions.find((key) => props[key]) ?? 'row'
}

export default direction
