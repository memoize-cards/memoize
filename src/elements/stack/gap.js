const gaps = [
  'quarck',
  'nano',
  'xs',
  'sm',
  'md',
  'lg',
  'huge',
  'giant'
]

function gap (props) {
  const gap = gaps.find((modifier) => props[modifier])
  return gap
    ? `var(--spacing_inset-${gap})`
    : '0'
}

export default gap
