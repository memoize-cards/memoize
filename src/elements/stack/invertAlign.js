function invertAlign (props) {
  return props.column ? 'justify-content' : 'align-items'
}

export default invertAlign
