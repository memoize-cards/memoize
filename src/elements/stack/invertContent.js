function invertContent (props) {
  return props.column ? 'align-items' : 'justify-content'
}

export default invertContent
