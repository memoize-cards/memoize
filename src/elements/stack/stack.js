import align from './align'
import content from './content'
import direction from './direction'
import gap from './gap'
import invertAlign from './invertAlign'
import invertContent from './invertContent'
import styled from '@standard/styled'
import tall from './tall'
import wide from './wide'

const stack = new Proxy({}, {
  get (_target, key) {
    return styled[key]`
      ${invertAlign}: ${align};
      display: flex;
      flex-direction: ${direction};
      gap: ${gap};
      height: ${tall};
      ${invertContent}: ${content};
      width: ${wide};
    `
  }
})

export default stack
