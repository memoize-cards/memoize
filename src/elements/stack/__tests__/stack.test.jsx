import h from '@standard/h'
import renderer from '@standard/renderer'
import stack from '@elements/stack'

describe('@elements/stack', () => {
  test('Should create an stack based on the tag name', async () => {
    const tree = await renderer(
      <stack.Div />
    )
    expect(tree).toMatchSnapshot()
  })
})
