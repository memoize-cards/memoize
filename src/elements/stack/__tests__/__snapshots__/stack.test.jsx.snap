// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@elements/stack Should create an stack based on the tag name 1`] = `
<div
  style="align-items: start; display: flex; flex-direction: row; gap: 0; height: auto; justify-content: start; width: inherit;"
/>
`;
